source/lac/trilinos_tpetra_vector.cc

```diff
@@ -0,0 +1,686 @@
+      Tpetra::Vector<Number, int, types::global_dof_index> source_vector(
+        tpetra_export.getSourceMap());
+
+      source_vector.template sync<Kokkos::HostSpace>();
+      auto x_2d = source_vector.template getLocalView<Kokkos::HostSpace>();
+      auto x_1d = Kokkos::subview(x_2d, Kokkos::ALL(), 0);
+      source_vector.template modify<Kokkos::HostSpace>();
+      const size_t localLength = source_vector.getLocalLength();
+      auto         values_it   = V.begin();
+      for (size_t k = 0; k < localLength; ++k)
+        x_1d(k) = *values_it++;
+      source_vector.template sync<
+        typename Tpetra::Vector<Number, int, types::global_dof_index>::
+          device_type::memory_space>();
+      if (operation == VectorOperation::insert)
+        vector->doExport(source_vector, tpetra_export, Tpetra::REPLACE);
+      else if (operation == VectorOperation::add)
+        vector->doExport(source_vector, tpetra_export, Tpetra::ADD);
+      else
+        AssertThrow(false, ExcNotImplemented());
+    }
+
+
+
+    template <typename Number>
+    Vector<Number> &
+    Vector<Number>::operator*=(const Number factor)
+    {
+      AssertIsFinite(factor);
+      vector->scale(factor);
+
+      return *this;
+    }
+
+
+
+    template <typename Number>
+    Vector<Number> &
+    Vector<Number>::operator/=(const Number factor)
+    {
+      AssertIsFinite(factor);
+      Assert(factor != Number(0.), ExcZero());
+      *this *= Number(1.) / factor;
+
+      return *this;
+    }
+
+
+
+    template <typename Number>
+    Vector<Number> &
+    Vector<Number>::operator+=(const VectorSpaceVector<Number> &V)
+    {
+      // Check that casting will work.
+      Assert(dynamic_cast<const Vector<Number> *>(&V) != nullptr,
+             ExcVectorTypeNotCompatible());
+
+      // Downcast V. If fails, throws an exception.
+      const Vector<Number> &down_V = dynamic_cast<const Vector<Number> &>(V);
+      // If the maps are the same we can Update right away.
+      if (vector->getMap()->isSameAs(*(down_V.trilinos_vector().getMap())))
+        {
+          vector->update(1., down_V.trilinos_vector(), 1.);
+        }
+      else
+        {
+          Assert(this->size() == down_V.size(),
+                 ExcDimensionMismatch(this->size(), down_V.size()));
+
+          // TODO: The code doesn't work as expected so we use a workaround.
+          /*Tpetra::Export<int, types::global_dof_index>
+          data_exchange(vector->getMap(), down_V.trilinos_vector().getMap());
+          vector->doExport(down_V.trilinos_vector(),
+                           data_exchange,
+                           Tpetra::ADD);*/
+
+          Tpetra::Vector<Number, int, types::global_dof_index> dummy(
+            vector->getMap(), false);
+          Tpetra::Import<int, types::global_dof_index> data_exchange(
+            dummy.getMap(), down_V.trilinos_vector().getMap());
+
+          dummy.doExport(down_V.trilinos_vector(),
+                         data_exchange,
```

Comment:
Is it normal to use an `Import` object with `doExpert`?

Reply:
No, a `doImport` also does the job here.